File: Sources/CoMaintainer/AutofixMonitor.swift
import Foundation
import CoMaintainerCore

// macOS side of the PR auto-fix monitor: fetch a snapshot of my open PRs (one GraphQL
// call via the `gh` CLI) so the shared AutofixDiff can decide what changed. The
// diffing lives in CoMaintainerCore; the spawn/track happens in Store. This file is
// just the GitHub read. The queries themselves live in `core/graphql/` with the other
// shared queries (single source of truth; a future Linux monitor reuses them as-is).
enum AutofixMonitor {
    /// One GraphQL search over my open, authored PRs in the target repo, with each
    /// PR's mergeability, review verdict, and review-thread resolution — everything the
    /// diff and the reconcilers need, in a single request (`core/graphql/monitor-prs`).
    static func fetchSnapshots(owner: String, repo: String, me: String) async throws -> [PRSnapshot] {
        let q = "repo:\(owner)/\(repo) author:\(me) is:pr is:open"
        let query = try CoreAssets.graphql("monitor-prs")
        let data = try await GH.run(["api", "graphql", "-f", "query=\(query)", "-f", "q=\(q)"])
        return try parse(data, me: me)
    }

    /// A PR that has requested my review, with the timestamps needed to decide whether I
    /// still owe a review — robustly, without depending on observing a "request removed"
    /// transition (which a re-request can slip past).
    struct ReviewRequest {
        let number: Int
        let title: String
        let url: String
        let author: String
        let authorAssociation: String // OWNER / MEMBER / COLLABORATOR / CONTRIBUTOR / NONE / …
        let files: [String]         // changed file paths (for skill/installer verdict gating)
        let requestedAt: String?    // latest "review requested from me" (ISO8601)
        let myLastReviewAt: String? // my latest review submission (ISO8601)

        /// I owe a review when I'm requested and that request is newer than my last review
        /// of this PR (ISO8601 strings compare chronologically). A fresh re-request (newer
        /// timestamp) re-qualifies even after I reviewed once.
        var oweReview: Bool {
            guard let r = requestedAt else { return true } // requested but no event detail → assume owed
            guard let m = myLastReviewAt else { return true }
            return r > m
        }
    }

    /// PRs that request MY review, with request/last-review timestamps (one GraphQL
    /// call, `core/graphql/review-requests`). `includeFiles` pulls each PR's changed
    /// paths for the verdict-withhold gate — a big chunk of the query's rate-limit
    /// cost, and pointless unless auto-approvals are on (verdicts are off by default),
    /// so the caller passes `false` then and the `@include` directive skips it.
    static func fetchReviewRequests(owner: String, repo: String, me: String,
                                    includeFiles: Bool = false) async throws -> [ReviewRequest] {
        let q = "repo:\(owner)/\(repo) review-requested:\(me) is:pr is:open"
        let query = try CoreAssets.graphql("review-requests")
        let data = try await GH.run(["api", "graphql", "-f", "query=\(query)", "-f", "q=\(q)",
                                     "-F", "withFiles=\(includeFiles)"])
        struct Resp: Decodable {
            let data: D
            struct D: Decodable { let search: S }
            struct S: Decodable { let nodes: [Node] }
            struct Node: Decodable {
                let number: Int?
                let title: String?
                let url: String?
                let author: Login?
                let authorAssociation: String?
                let files: FilesConn?
                let timelineItems: TL?
                let reviews: RVs?
            }
            struct Login: Decodable { let login: String? }
            struct FilesConn: Decodable { let nodes: [FileNode] }
            struct FileNode: Decodable { let path: String? }
            struct TL: Decodable { let nodes: [Ev] }
            struct Ev: Decodable { let createdAt: String?; let requestedReviewer: Login? }
            struct RVs: Decodable { let nodes: [RV] }
            struct RV: Decodable { let author: Login?; let submittedAt: String? }
        }
        let r = try JSONDecoder().decode(Resp.self, from: data)
        let lower = me.lowercased()
        return r.data.search.nodes.compactMap { n in
            guard let number = n.number else { return nil }
            let reqAt = (n.timelineItems?.nodes ?? [])
                .filter { $0.requestedReviewer?.login?.lowercased() == lower }
                .compactMap { $0.createdAt }.max()
            let myReviewAt = (n.reviews?.nodes ?? [])
                .filter { $0.author?.login?.lowercased() == lower }
                .compactMap { $0.submittedAt }.max()
            return ReviewRequest(number: number, title: n.title ?? "", url: n.url ?? "",
                                 author: n.author?.login ?? "",
                                 authorAssociation: n.authorAssociation ?? "NONE",
                                 files: (n.files?.nodes ?? []).compactMap { $0.path },
                                 requestedAt: reqAt, myLastReviewAt: myReviewAt)
        }
    }

    /// Decode the GraphQL search response into snapshots. Non-PR search nodes (no
    /// `number`) are skipped. `me` distinguishes threads I still owe a reply on (last
    /// comment isn't mine) from ones waiting on the reviewer — via the shared
    /// `ThreadTriage.owed` rule (same one the Tool 6 list uses).
    static func parse(_ data: Data, me: String = "") throws -> [PRSnapshot] {
        struct Resp: Decodable {
            let data: DataField
            struct DataField: Decodable { let search: Search }
            struct Search: Decodable { let nodes: [Node] }
            struct Node: Decodable {
                let number: Int?
                let title: String?
                let url: String?
                let isDraft: Bool?
                let mergeable: String?
                let reviewDecision: String?
                let reviewThreads: Threads?
            }
            struct Author: Decodable { let login: String? }
            struct Threads: Decodable { let nodes: [Thread] }
            struct Thread: Decodable {
                let isResolved: Bool
                let viewerCanResolve: Bool?
                let comments: Comments?
                struct Comments: Decodable { let nodes: [Comment] }
                struct Comment: Decodable { let author: Author? }
            }
        }
        let r = try JSONDecoder().decode(Resp.self, from: data)
        return r.data.search.nodes.compactMap { n in
            guard let number = n.number else { return nil }
            let threads = n.reviewThreads?.nodes ?? []
            let unresolved = threads.filter { !$0.isResolved }.count
            let iOwe = threads.filter { t in
                ThreadTriage.owed(isResolved: t.isResolved, viewerCanResolve: t.viewerCanResolve,
                                  lastCommentAuthor: t.comments?.nodes.last?.author?.login, me: me)
            }.count
            return PRSnapshot(
                number: number,
                title: n.title ?? "",
                url: n.url ?? "",
                isDraft: n.isDraft ?? false,
                mergeable: n.mergeable ?? "UNKNOWN",
                reviewDecision: n.reviewDecision ?? "",
                threadsUnresolved: unresolved,
                threadsIOwe: iOwe)
        }
    }
}

File: linux/co_maintainer/autofixmonitor.py
"""GitHub reads for the PR auto-fix monitor — the Linux port of AutofixMonitor.swift.

Two GraphQL searches over the `gh` CLI, decoded into the pure types in
:mod:`autofix`. The queries themselves are the shared ones in ``core/graphql/``
(single source of truth with the macOS monitor), driven by a search ``$q``
qualifier rather than the owner/name form that :func:`gh.graphql` wraps — so the
requests are issued directly here.
"""

from __future__ import annotations

import json

from . import core, gh
from .autofix import PRSnapshot, ReviewRequest


def _owed_thread(thread: dict, me: str) -> bool:
    """A review thread I still owe a reply on (mirrors ThreadTriage.owed): unresolved,
    resolvable by me, and whose most-recent comment isn't mine."""
    if thread.get("isResolved"):
        return False
    vcr = thread.get("viewerCanResolve")
    if vcr is None:
        vcr = True
    if not vcr:
        return False
    comments = (thread.get("comments") or {}).get("nodes") or []
    last = comments[-1] if comments else None
    login = ((last or {}).get("author") or {}).get("login") if last else None
    return (login or "").lower() != me.lower()


def _parse_snapshots(env: dict, me: str) -> list[PRSnapshot]:
    nodes = ((env.get("data") or {}).get("search") or {}).get("nodes") or []
    out: list[PRSnapshot] = []
    for n in nodes:
        if not n:
            continue
        number = n.get("number")
        if number is None:  # non-PR search node
            continue
        threads = (n.get("reviewThreads") or {}).get("nodes") or []
        unresolved = sum(1 for t in threads if t and not t.get("isResolved"))
        i_owe = sum(1 for t in threads if t and _owed_thread(t, me))
        out.append(
            PRSnapshot(
                number=number,
                title=n.get("title") or "",
                url=n.get("url") or "",
                is_draft=bool(n.get("isDraft")),
                mergeable=n.get("mergeable") or "UNKNOWN",
                review_decision=n.get("reviewDecision") or "",
                threads_unresolved=unresolved,
                threads_i_owe=i_owe,
            )
        )
    return out


def fetch_snapshots(owner: str, repo: str, me: str) -> list[PRSnapshot]:
    """One GraphQL search over my open, authored PRs (``core/graphql/monitor-prs``)."""
    q = f"repo:{owner}/{repo} author:{me} is:pr is:open"
    query = core.read_graphql("monitor-prs")
    data = gh.run(["api", "graphql", "-f", f"query={query}", "-f", f"q={q}"])
    return _parse_snapshots(json.loads(data), me)


def _parse_review_requests(env: dict, me: str) -> list[ReviewRequest]:
    nodes = ((env.get("data") or {}).get("search") or {}).get("nodes") or []
    lower = me.lower()
    out: list[ReviewRequest] = []
    for n in nodes:
        if not n:
            continue
        number = n.get("number")
        if number is None:
            continue
        # Latest "review requested from me" event.
        req_times = [
            ev.get("createdAt")
            for ev in ((n.get("timelineItems") or {}).get("nodes") or [])
            if ev
            and ((ev.get("requestedReviewer") or {}).get("login") or "").lower() == lower
            and ev.get("createdAt")
        ]
        # My latest review submission on this PR.
        my_times = [
            rv.get("submittedAt")
            for rv in ((n.get("reviews") or {}).get("nodes") or [])
            if rv
            and ((rv.get("author") or {}).get("login") or "").lower() == lower
            and rv.get("submittedAt")
        ]
        out.append(
            ReviewRequest(
                number=number,
                title=n.get("title") or "",
                url=n.get("url") or "",
                author=((n.get("author") or {}).get("login")) or "",
                author_association=n.get("authorAssociation") or "NONE",
                files=[
                    f.get("path")
                    for f in ((n.get("files") or {}).get("nodes") or [])
                    if f and f.get("path")
                ],
                requested_at=max(req_times) if req_times else None,
                my_last_review_at=max(my_times) if my_times else None,
            )
        )
    return out


def fetch_review_requests(
    owner: str, repo: str, me: str, include_files: bool = False
) -> list[ReviewRequest]:
    """PRs that request MY review, with request/last-review timestamps
    (``core/graphql/review-requests``). ``include_files`` pulls each PR's changed
    paths for the verdict-withhold gate — a big chunk of the rate-limit cost, so
    the caller passes ``False`` unless auto-approvals are on."""
    q = f"repo:{owner}/{repo} review-requested:{me} is:pr is:open"
    query = core.read_graphql("review-requests")
    data = gh.run(
        [
            "api",
            "graphql",
            "-f",
            f"query={query}",
            "-f",
            f"q={q}",
            "-F",
            f"withFiles={'true' if include_files else 'false'}",
        ]
    )
    return _parse_review_requests(json.loads(data), me)
